airflow/jobs/scheduler_job.py

```diff
@@ -1568,3 +1590,21 @@ def _cleanup_stale_dags(self, session: Session = NEW_SESSION) -> None:
             dag.is_active = False
             SerializedDagModel.remove_dag(dag_id=dag.dag_id, session=session)
         session.flush()
+
+    def notify_dagrun_state_changed(self, dag_run: DagRun, msg: str = ""):
+        if not self.enabled_dagrun_listener or not get_listener_manager().has_scheduler_listeners:
+            return
+
+        if dag_run.state == DagRunState.RUNNING:
+            self._notification_threadpool.submit(  # type: ignore[union-attr]
+                get_listener_manager().hook.on_dag_run_start, dag_run=dag_run, msg=msg
+            )
+        elif dag_run.state == DagRunState.SUCCESS:
+            self._notification_threadpool.submit(  # type: ignore[union-attr]
+                get_listener_manager().hook.on_dag_run_success, dag_run=dag_run, msg=msg
+            )
+        elif dag_run.state == DagRunState.FAILED:
+            self._notification_threadpool.submit(  # type: ignore[union-attr]
+                get_listener_manager().hook.on_dag_run_failure, dag_run=dag_run, msg=msg
+            )
```

Comment:
I'm not sure the threadpool belongs in here.

If we make that the _plugin's_ responsibility then

a) I don't think forcing a threadpool (of a fixed size, but that could be config driven) on users of this hook is required;
b) We probably don't need `has_scheduler_listeners` anymore; 
b) This PR becomes a lot smaller.

Reply:
Removed threadpool.